### PR TITLE
Test: Frontend 63

### LINKED ISSUE #63: Frontend: Add Unit Tests
Add unit tests for the frontend components to ensure they function correctly and handle edge cases.

### PR DESCRIPTION
**Pull request by GregoryNavasarkian**
### Closes #63 

This pull request introduces several changes to the client-side codebase, focusing on adding testing capabilities, updating dependencies, and improving component functionality. The most important changes include adding test scripts and configurations, updating dependencies, and enhancing the UI components with better accessibility and testability.

### Testing Enhancements:
* [`client/package.json`](diffhunk://#diff-1846122c2c83a486a3693f7966aa522c34cf489f674185c4da0d9221683fd81fR10-R12): Added test scripts (`test`, `test:watch`, `test:coverage`) and necessary testing dependencies (`@testing-library/jest-dom`, `@testing-library/react`, `@testing-library/user-event`, `jsdom`, `vitest`, `@vitest/coverage-v8`). [[1]](diffhunk://#diff-1846122c2c83a486a3693f7966aa522c34cf489f674185c4da0d9221683fd81fR10-R12) [[2]](diffhunk://#diff-1846122c2c83a486a3693f7966aa522c34cf489f674185c4da0d9221683fd81fL19-R45)
* [`client/src/setupTests.ts`](diffhunk://#diff-222b4809575578b31d53e05644b0a27786bc4a7537992a1d75e1031211e30e2fR1): Imported `@testing-library/jest-dom` to extend Jest matchers for the DOM.
* `client/tests/App.test.tsx`, `client/tests/components/Insights.test.tsx`, `client/tests/components/MainNavbar.test.tsx`, `client/tests/components/RecordingForm.test.tsx`: Added comprehensive test cases for various components to ensure correct rendering and functionality. [[1]](diffhunk://#diff-d5b0f2925912fa69cf926d1b44f8c79036c5b71b50f88263826a40fe04cbaf2cR1-R23) [[2]](diffhunk://#diff-3aec674aded074922b7637de956f78b2851fef6774e7791eca725ec6f755e4d4R1-R226) [[3]](diffhunk://#diff-d8ccfda21775007724ce8e1f8a10f40a059f8bbd842dec4562120ab31e369090R1-R18) [[4]](diffhunk://#diff-07536c92f1511ad53773d1f3dda7784598cec26a1d611d60971e315ed7e8901dR1-R254)

### Dependency Updates:
* [`client/package.json`](diffhunk://#diff-1846122c2c83a486a3693f7966aa522c34cf489f674185c4da0d9221683fd81fL19-R45): Updated `react-datepicker` to version `8.2.1` and added new dependencies for testing and coverage.

### UI Enhancements:
* [`client/src/components/ExerciseLogList.tsx`](diffhunk://#diff-ae5260681aa437085bbbb8e45efb8adc0bf61dd70da2a8f6b3717b3ece755662L165-R171): Enhanced the loading spinner with `data-testid` and `role` attributes for better accessibility and testability.
* [`client/src/components/Insights.tsx`](diffhunk://#diff-6b16ef16599a7504c4778eaa0437aa2340f6a4ac5a1332fd60c7a84e67e9d840L164-R164): Added `role` attribute to the loading spinner and exported `CustomTooltip` for testing purposes. [[1]](diffhunk://#diff-6b16ef16599a7504c4778eaa0437aa2340f6a4ac5a1332fd60c7a84e67e9d840L164-R164) [[2]](diffhunk://#diff-6b16ef16599a7504c4778eaa0437aa2340f6a4ac5a1332fd60c7a84e67e9d840R334)

### Workflow Improvements:
* [`.github/workflows/workflow.yml`](diffhunk://#diff-126bf89616b7daa3d14ebc882ad18666aaf1c3dae888c4ba306a66ec80758bc1R62-R66): Added a new job to test the client-side code, ensuring that client tests are run as part of the CI pipeline.

### Gitignore Update:
* [`client/.gitignore`](diffhunk://#diff-f8fc230f40372bc0cc3003a0913de478a6b6cd220762e6896942c383730974c8R26-R27): Added `coverage` to ignore coverage reports generated by tests.

### CODE REVIEW

File: client/src/components/ExerciseLogList.tsx [hunk 1]
```diff
@@ -188,315 +194,315 @@ const ExerciseLogList: React.FC = () => {
...
+                            {editingLogId !== log._id && (
                               <Button
-                                variant="secondary"
+                                variant="danger"
                                 className="mt-2"
-                                onClick={() => handleCancel(log._id)}
+                                onClick={() => handleDelete(log._id)}
                               >
-                                Cancel
+                                Delete
```
Comment (JamesWang0210) [inline client/src/components/ExerciseLogList.tsx:494]: I bet these don't change the functionality itself right? you just switched to different bootstrap button types?

File: client/tests/components/ExerciseLogList.test.tsx [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+      deleteExerciseLog: mockDeleteExerciseLog,
+    });
+  });
+
+  it("renders logs after loading", async () => {
+    render(
+      <MemoryRouter>
+        <ExerciseLogList />
+      </MemoryRouter>,
+    );
```
Comment (JamesWang0210) [inline client/tests/components/ExerciseLogList.test.tsx:60]: not a merge blocker. but see if you want to create a util to call this render in order to save some lines of code. I can see the same render in multiple test suites. Unless there are changes of props to be passed into log list that you want to test. But still that is doable with a util function. You can pass in diff props that you want to test from my perspective. up to you.

File: client/tests/components/ExerciseLogList.test.tsx [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+          exercises: [
+            {
+              exercise_name: "Running",
+              type: "Cardio",
+              time: 45,
+            },
+          ],
+        },
+        dateTime: "2025-03-09T09:30:00.000Z",
+      },
```
Comment (JamesWang0210) [inline client/tests/components/ExerciseLogList.test.tsx:417]: nit: you can create a const for this as this is reused

File: client/tests/components/Insights.test.tsx [hunk 4]
```diff
@@ -0,0 +1,226 @@
+import { fireEvent, render, screen } from "@testing-library/react";
+import Insights, { CustomTooltip } from "../../src/components/Insights";
+import useExerciseLogs from "../../src/hooks/useExerciseLogs";
+
+global.ResizeObserver = class ResizeObserver {
+  observe() {}
+  unobserve() {}
+  disconnect() {}
+};
```
Comment (JamesWang0210) [inline client/tests/components/Insights.test.tsx:9]: what is this for

File: client/tests/components/RecordingForm.test.tsx [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+    vi.stubGlobal("navigator", {
+      mediaDevices: {
+        getUserMedia: vi.fn(() => Promise.resolve({})),
+      },
+    });
+
+    await act(async () => {
+      render(<RecordingForm />);
+    });
```
Comment (JamesWang0210) [inline client/tests/components/RecordingForm.test.tsx:64]: help me recall. is await act being deprecated since react 18? or only for a specific use case? i vaguely remember it

File: client/tests/components/RecordingForm.test.tsx [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+global.fetch = vi.fn(() =>
+  Promise.resolve({
+    json: () =>
+      Promise.resolve({
+        success: true,
+        transcript: "Test transcript",
+        result: [],
+      }),
+  }),
+) as any;
```
Comment (JamesWang0210) [inline client/tests/components/RecordingForm.test.tsx:15]: do we actually want to mock fetch in this way? This will force all fetch to return the same response correct? i think for our project its fine since in the following test one fetch and only one fetch is made every time. Just want to make sure we don't have multiple fetches involved in one test

File: client/src/components/ExerciseLogList.tsx (hunk 1 above)
Comment (GregoryNavasarkian) [reply client/src/components/ExerciseLogList.tsx:494]: I think its a formatting change. The functionality is the same as far as I can tell.

File: client/tests/components/ExerciseLogList.test.tsx (hunk 2 above)
Comment (GregoryNavasarkian) [reply client/tests/components/ExerciseLogList.test.tsx:60]: I think we can change things like this in a future refactor PR. For now it works so I just left it. Getting the tests to work was a pain since I am writing after the code was already written.

File: client/tests/components/ExerciseLogList.test.tsx (hunk 3 above)
Comment (GregoryNavasarkian) [reply client/tests/components/ExerciseLogList.test.tsx:417]: Again good for a refactor.

File: client/tests/components/Insights.test.tsx (hunk 4 above)
Comment (GregoryNavasarkian) [reply client/tests/components/Insights.test.tsx:9]: No idea, but it did not work without it. lol

File: client/tests/components/RecordingForm.test.tsx (hunk 5 above)
Comment (GregoryNavasarkian) [reply client/tests/components/RecordingForm.test.tsx:64]: Act from 'react-dom' is deprecated, it was replaced with act from 'react'. Its not necessary it just prevents warnings in test output in the terminal.

File: client/tests/components/RecordingForm.test.tsx (hunk 6 above)
Comment (GregoryNavasarkian) [reply client/tests/components/RecordingForm.test.tsx:15]: I think its fine. I am not great when it comes to testing so this is the best I could come up with. Feel free to clone the branch and make changes if you want.

Comment (GregoryNavasarkian) [reply client/tests/components/RecordingForm.test.tsx:15]: It's just testing the functionality and if the logic is working. Since the backend is tested separately, there was no point in mocking proper requests and responses.

File: client/tests/components/ExerciseLogList.test.tsx (hunk 2 above)
Comment (JamesWang0210) [reply client/tests/components/ExerciseLogList.test.tsx:60]: sure. not a blocker like what I said

File: client/tests/components/Insights.test.tsx (hunk 4 above)
Comment (JamesWang0210) [reply client/tests/components/Insights.test.tsx:9]: lol gotcha. we may not need it. feel free to clean it out later on.

File: client/tests/components/RecordingForm.test.tsx (hunk 5 above)
Comment (JamesWang0210) [reply client/tests/components/RecordingForm.test.tsx:64]: oh yes I recall that warning now. good

File: client/tests/components/RecordingForm.test.tsx (hunk 6 above)
Comment (JamesWang0210) [reply client/tests/components/RecordingForm.test.tsx:15]: yeah thats right. there could be a better way. but this is good enough for us for now. Yeah I can give it a try.

**Review by JamesWang0210 — APPROVED** [review]

**Review by jtp03a — APPROVED** [review]
